### PR TITLE
[Data] Support using concurrent actors for `ActorPool`

### PR DESCRIPTION
**Pull request by amogkam**
Support using concurrent actors for `ActorPool`. We do this by gating the user UDF in a separate threadpool of max size 1.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — COMMENTED** [review]
A couple questions:
1. Should we always use this thread-pool code path regardless of the setting?
2. Do we have any segfault issues right now?
3. How does this relate to the thread-based pipelining code you have for accelerated prefetch? It seems like we could also use that pipeline verbatim, with each actor thread feeding into the pipeline and awaiting the output at the end.

**Comment by amogkam** [comment]
1. We can. Are you suggesting for code simplicity?
2. This approach doesn't run into segfault
3. Hmm I don't think that can very easily be plugged in as is. I think the pattern is different right? For prefetching there is a single producer with multiple threads being used for the computation. Here, there are multiple producers, with a single thread being used for computation.

**Comment by amogkam** [comment]
Made the change for point 1

File: python/ray/data/tests/test_dataset_map.py [hunk 1]
```diff
@@ -144,6 +145,31 @@ def __call__(self, x):
...
+        def __call__(self, x):
+            thread_id = threading.get_ident()
+            assert threading.current_thread() is not threading.main_thread()
+            return [thread_id]
+
+    thread_ids = ds.map_batches(
+        StatefulFn, compute="actors", max_concurrency=2
+    ).take_all()
+    # Make sure user's UDF is not running concurrently.
+    assert len(set(thread_ids)) == 1
```
Comment (amogkam) [inline python/ray/data/tests/test_dataset_map.py:163]: test here @ericl

**Review by ericl — APPROVED** [review]

**Review by c21 — APPROVED** [review]
should we update documentation as well? so users are aware that they can use the concurrent actors by setting `max_concurrency` in `map_batches`?

**Pull request merged**